Tutorial example for computing the radiation pattern of axisymmetric and nonaxisymmetric linearly polarized dipoles in cylindrical coordinates (#2950)
NanoComp/meep: doc/docs/Python_Tutorials/Near_to_Far_Field_Spectra.md

When these two conditions are not met as in the example below involving a small `dpad` and large `d2`, the error from the finite truncation and numerical dispersion can be large and therefore result in a significant mismatch between the far fields computed using the near-to-far field transformation versus the actual DFT fields at the same location.

![Comparison of the far fields from the near-to-far field transformation dominated by errors and the DFT fields at the same location for a holey-waveguide cavity.](../images/farfields_vs_DFTfields_holeycavity_mismatch.png#center)

Radiation Pattern of an Antenna in Cylindrical Coordinates
----------------------------------------------------------

> Earlier we showed how to compute the [radiation pattern of an antenna with linear polarization in vacuum](#radiation-pattern-of-an-antenna) using a simulation in 2D Cartesian coordinates. The same calculation can also be performed using [cylindrical coordinates](../Exploiting_Symmetry.md#cylindrical-symmetry). We will compute the radiation pattern for two different cases in which the dipole is (1) axisymmetric (i.e., at $r = 0$) or (2) nonaxisymmetric (i.e., at $r > 0$). The radiation pattern is validated using the analytic result from antenna theory.

Reviewer: ```suggestion
Earlier we showed how to compute the [radiation pattern of an antenna with linear polarization in vacuum](#radiation-pattern-of-an-antenna) using a simulation in 2D Cartesian coordinates. The same calculation can also be performed using [cylindrical coordinates](../Exploiting_Symmetry.md#cylindrical-symmetry). We will compute the radiation pattern for two different cases in which the dipole is (1) on-axis (i.e., at $r = 0$) or (2) off-axis (i.e., at $r > 0$). The radiation pattern is validated using the analytic result from antenna theory in vacuum—of course, in a homogeneous medium, the on- and off-axis cases are equivalent to observers at infinity.
```

> In this example, the radiation pattern is computed for $\phi = 0$ (i.e., the $rz$ or $xz$ plane). Note that the radiation pattern for an $\hat{x}$ or $\hat{y}$ polarized dipole is *nonaxisymmetric*. This means that the radiation pattern depends on the choice of $\phi$. (This is different than the computation of the [extraction efficiency for nonaxisymmetric dipoles](Cylindrical_Coordinates.md#nonaxisymmetric-dipole-sources) which involves an integration over $\phi$ resulting in a radiation pattern that is axisymmetric.)

Reviewer: ```suggestion
In this example, the radiation pattern is computed for $\phi = 0$ (i.e., the $rz$ or $xz$ plane). Note that the radiation pattern for an $\hat{x}$ or $\hat{y}$ polarized dipole is *nonaxisymmetric* even if it is on-axis. This means that the radiation pattern depends on the choice of $\phi$. (This is different than the computation of the [extraction efficiency for nonaxisymmetric dipoles](Cylindrical_Coordinates.md#nonaxisymmetric-dipole-sources) which involves an integration over $\phi$ resulting in a radiation pattern that is axisymmetric.)
```

> For (1), there are two dipole configurations: $E_x$ and $E_z$. An $E_z$ dipole is positioned at $r = 0$ with $m = 0$. This involves a single simulation. An $E_x$ dipole at $r = 0$, however, involves the superposition of left- and right-circularly polarized dipoles ($E_r \pm iE_\phi$) as described in [Tutorial/Scattering Cross Section of a Finite Dielectric Cylinder](Cylindrical_Coordinates.md#scattering-cross-section-of-a-finite-dielectric-cylinder). This requires *two* simulations. The computation of the radiation pattern of an $E_x$ dipole at $r = 0$ is different from the [computation of its extraction efficiency](Local_Density_of_States.md#extraction-efficiency-of-a-light-emitting-diode-led) which involves a *single* $E_r$ source with either $m = +1$ or $m = -1$. This is because extraction efficiency involves integration over $\phi$, in which case the $m = \pm 1$ cross terms cancel and you simply get identical integrated powers from $m = \pm 1$.

Reviewer: ```suggestion
For (1), there are two dipole configurations: $E_x$ and $E_z$. An $E_z$ dipole is positioned at $r = 0$ with $m = 0$. This involves a single simulation. An $E_x$ dipole at $r = 0$, however, involves the superposition of left- and right-circularly polarized dipoles ($E_r \pm iE_\phi$) as described in [Tutorial/Scattering Cross Section of a Finite Dielectric Cylinder](Cylindrical_Coordinates.md#scattering-cross-section-of-a-finite-dielectric-cylinder). This requires *two* simulations at $m = \pm 1$ (although it is possible to combine these into a single $m=+1$ simulation at $\pm \omega$ as described below). The computation of the radiation pattern of an $E_x$ dipole at $r = 0$ is different from the [computation of its extraction efficiency](Local_Density_of_States.md#extraction-efficiency-of-a-light-emitting-diode-led) which involves a *single* $E_r$ source with either $m = +1$ or $m = -1$. This is because extraction efficiency involves integration over $\phi$, in which case the $m = \pm 1$ cross terms cancel and you simply get identical integrated powers from $m = \pm 1$.
```

> For (1), there are two dipole configurations: $E_x$ and $E_z$. An $E_z$ dipole is positioned at $r = 0$ with $m = 0$. This involves a single simulation. An $E_x$ dipole at $r = 0$, however, involves the superposition of left- and right-circularly polarized dipoles ($E_r \pm iE_\phi$) as described in [Tutorial/Scattering Cross Section of a Finite Dielectric Cylinder](Cylindrical_Coordinates.md#scattering-cross-section-of-a-finite-dielectric-cylinder). This requires *two* simulations. The computation of the radiation pattern of an $E_x$ dipole at $r = 0$ is different from the [computation of its extraction efficiency](Local_Density_of_States.md#extraction-efficiency-of-a-light-emitting-diode-led) which involves a *single* $E_r$ source with either $m = +1$ or $m = -1$. This is because the latter calculation involves a circularly polarized source which emits exactly half the power as a linearly polarized source even though their radiation patterns are different: $\frac{1}{2}(1 + \cos^2\theta)$ vs. $\cos^2\theta$.

Reviewer: ```suggestion
For (1), there are two dipole configurations: $E_x$ and $E_z$. An $E_z$ dipole is positioned at $r = 0$ with $m = 0$. This involves a single simulation. An $E_x$ dipole at $r = 0$, however, involves the superposition of left- and right-circularly polarized dipoles ($E_r \pm iE_\phi$) as described in [Tutorial/Scattering Cross Section of a Finite Dielectric Cylinder](Cylindrical_Coordinates.md#scattering-cross-section-of-a-finite-dielectric-cylinder). This requires *two* simulations. The computation of the radiation pattern of an $E_x$ dipole at $r = 0$ is different from the [computation of its extraction efficiency](Local_Density_of_States.md#extraction-efficiency-of-a-light-emitting-diode-led) which involves a *single* $E_r$ source with either $m = +1$ or $m = -1$.  This is because extraction efficiency involves integration over $\phi$, in which case the $m = \pm 1$ cross terms cancel and you simply get identical integrated powers from $m = \pm 1$.
```

> For (1), there are two dipole configurations: $E_x$ and $E_z$. An $E_z$ dipole is positioned at $r = 0$ with $m = 0$. This involves a single simulation. An $E_x$ dipole at $r = 0$, however, involves the superposition of left- and right-circularly polarized dipoles ($E_r \pm iE_\phi$) as described in [Tutorial/Scattering Cross Section of a Finite Dielectric Cylinder](Cylindrical_Coordinates.md#scattering-cross-section-of-a-finite-dielectric-cylinder). This requires *two* simulations at $m = \pm 1$ (although it is possible to combine these into a single $m=+1$ simulation at $\pm \omega$ as described below). The computation of the radiation pattern of an $E_x$ dipole at $r = 0$ is different from the [computation of its extraction efficiency](Local_Density_of_States.md#extraction-efficiency-of-a-light-emitting-diode-led) which involves a *single* $E_r$ source with either $m = +1$ or $m = -1$. This is because extraction efficiency involves integration over $\phi$, in which case the $m = \pm 1$ cross terms cancel and you simply get identical integrated powers from $m = \pm 1$.

Reviewer: ```suggestion
For on-axis (1), there are two dipole configurations: $E_x$ and $E_z$. An $E_z$ dipole is positioned at $r = 0$ with $m = 0$. This involves a single simulation. An $E_x$ dipole at $r = 0$, however, involves the superposition of left- and right-circularly polarized dipoles ($E_r \pm iE_\phi$) as described in [Tutorial/Scattering Cross Section of a Finite Dielectric Cylinder](Cylindrical_Coordinates.md#scattering-cross-section-of-a-finite-dielectric-cylinder). This requires *two* simulations at $m = \pm 1$ (although it is possible to combine these into a single $m=+1$ simulation at $\pm \omega$ as described below). The computation of the radiation pattern of an $E_x$ dipole at $r = 0$ is different from the [computation of its extraction efficiency](Local_Density_of_States.md#extraction-efficiency-of-a-light-emitting-diode-led) which involves a *single* $E_r$ source with either $m = +1$ or $m = -1$. This is because extraction efficiency involves integration over $\phi$, in which case the $m = \pm 1$ cross terms cancel and you simply get identical integrated powers from $m = \pm 1$.
```

> For (2), an $E_x$ (or equivalently an $E_r$) dipole positioned at $r > 0$ requires a [Fourier-series expansion of the fields](Cylindrical_Coordinates.md#nonaxisymmetric-dipole-sources) from an $E_r$ "ring" current source with azimuthal dependence $e^{im\phi}$. The Fourier-series expansion involves nonnegative $m$ values because the $-m$ fields can be obtained directly from the $+m$ fields using the relation $\vec{E}_{m}(r, z)e^{im\phi-i\omega t} = \left[\vec{E}_{-m}(r, z)e^{-im\phi+i\omega t} \right]^*$. There is a similar relation for $\vec{H}_{m}$. This involves taking the $\vec{E}_{m}$ fields at $-\omega$ (not $\omega$!) and conjugating them to get the $\vec{E}_{-m}$ fields at $\omega$. Implementing this in a simulation involves two aspects: (1) two sources with frequencies $\omega$ and $-\omega$ and (2) two DFT field monitors at $\omega$ and $-\omega$. (1) is necessary because it ensures the source current $\mathcal{J}$ is purely real and therefore does not change under a complex conjugation. We will compute the radiation pattern for $E_x$ and $E_y$ dipoles at $r = 0.1$ μm.

Reviewer: ```suggestion
For off-axis (2), an $E_x$ (or equivalently an $E_r$) dipole positioned at $r > 0$ requires a [Fourier-series expansion of the fields](Cylindrical_Coordinates.md#nonaxisymmetric-dipole-sources) from an $E_r$ "ring" current source with azimuthal dependence $e^{im\phi}$. As an optimization, our Fourier-series expansion involves only nonnegative $m$ values because the $-m$ fields can be obtained directly from the $+m$ fields using the relation $\vec{E}_{m}(r, z)e^{im\phi-i\omega t} = \left[\vec{E}_{-m}(r, z)e^{-im\phi+i\omega t} \right]^*$. There is a similar relation for $\vec{H}_{m}$. This involves taking the $\vec{E}_{m}$ fields at $-\omega$ (not $\omega$!) and conjugating them to get the $\vec{E}_{-m}$ fields at $\omega$. Implementing this in a simulation involves two aspects: (1) two sources with frequencies $\omega$ and $-\omega$ and (2) two DFT field monitors at $\omega$ and $-\omega$. (1) is necessary because it ensures the source current $\mathcal{J}$ is purely real and therefore does not change under a complex conjugation. We will compute the radiation pattern for $E_x$ and $E_y$ dipoles at $r = 0.1$ μm.
```
